contracts/src/finance/vesting_wallet.rs

```diff
@@ -0,0 +1,621 @@
+    /// * `token` - Address of the token being released.
+    ///
+    /// # Errors
+    ///
+    /// If the `token` address is not a contract, then the error
+    /// [`Error::InvalidToken`] is returned.
+    /// If the contract fails to execute the call, then the error
+    /// [`Error::ReleaseTokenFailed`] is returned.
+    ///
+    /// # Events
+    ///
+    /// Emits an [`ERC20Released`] event.
+    ///
+    /// # Panics
+    ///
+    /// If total allocation exceeds `U256::MAX`.
+    /// If scaled total allocation (mid calculation) exceeds `U256::MAX`.
+    #[selector(name = "release")]
+    fn release_erc20(&mut self, token: Address) -> Result<(), Self::Error>;
+
+    /// Calculates the amount of Ether that has already vested. Default
+    /// implementation is a linear vesting curve.
+    ///
+    /// # Arguments
+    ///
+    /// * `&self` - Read access to the contract's state.
+    /// * `timestamp` - Point in time for which to check the vested amount.
+    ///
+    /// # Panics
+    ///
+    /// If total allocation exceeds `U256::MAX`.
+    /// If scaled total allocation (mid calculation) exceeds `U256::MAX`.
+    #[selector(name = "vestedAmount")]
+    fn vested_amount_eth(&self, timestamp: u64) -> U256;
+
+    /// Calculates the amount of tokens that has already vested. Default
+    /// implementation is a linear vesting curve.
+    ///
+    /// # Arguments
+    ///
+    /// * `&mut self` - Write access to the contract's state.
+    /// * `token` - Address of the token being released.
+    /// * `timestamp` - Point in time for which to check the vested amount.
+    ///
+    /// # Errors
+    ///
+    /// If the `token` address is not a contract, then the error
+    /// [`Error::InvalidToken`] is returned.
+    ///
+    /// # Panics
+    ///
+    /// If total allocation exceeds `U256::MAX`.
+    /// If scaled total allocation (mid calculation) exceeds `U256::MAX`.
+    #[selector(name = "vestedAmount")]
+    fn vested_amount_erc20(
+        &mut self,
+        token: Address,
+        timestamp: u64,
+    ) -> Result<U256, Self::Error>;
+}
+
+#[public]
+impl IVestingWallet for VestingWallet {
+    type Error = Error;
+
+    fn owner(&self) -> Address {
+        self.ownable.owner()
+    }
+
+    fn transfer_ownership(
+        &mut self,
+        new_owner: Address,
+    ) -> Result<(), Self::Error> {
+        Ok(self.ownable.transfer_ownership(new_owner)?)
+    }
+
+    fn renounce_ownership(&mut self) -> Result<(), Self::Error> {
+        Ok(self.ownable.renounce_ownership()?)
+    }
+
+    #[payable]
+    fn receive_ether(&self) {}
+
+    fn start(&self) -> U256 {
+        U256::from(self._start.get())
+    }
+
+    fn duration(&self) -> U256 {
+        U256::from(self._duration.get())
+    }
+
+    fn end(&self) -> U256 {
+        // SAFETY: both `start` and `duration` are stored as u64,
+        // so they cannot exceed `U256::MAX`
+        self.start() + self.duration()
```

Comment:
```suggestion
        self.start().checked_add(self.duration()).expect("should not panic")
```

Reply:
But seems panic shouldn't be possible. Make sense to omit the check, isn't it?

Reply:
Right now it is not possible, but what if the contract would use `U256` for `start` and `duration` in the future?

Reply:
It will look like a mistake then. U64 timestamp should be more then enough for the current century :)

Reply:
I see that at solidity it was done, to remove unnecessary checks. And improve gas usage. Idk really how is it related to our case. But yeah, I don't mind to copy their approach for now.
Besides @0xNeshi added edge case to the `reads_max_end` test case